raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,321 @@
+
+
+        // Call updateTransfer in the corresponding TokenNetwork
+        token_network.updateNonClosingBalanceProof(
+            closing_participant,
+            non_closing_participant,
+            balance_hash,
+            nonce,
+            additional_hash,
+            closing_signature,
+            non_closing_signature
+        );
+
+        emit NewBalanceProofReceived(
+            reward_proof_signature,
+            nonce,
+            msg.sender,
+            reward_sender_address
+        );
+    }
+
+    /// @notice Called after a monitored channel is settled in order for MS to claim the reward
+    /// Can be called once per settled channel by everyone on behalf of MS
+    /// @param token_network_address Address of the Token Network in which the channel
+    /// @param reward_amount Amount to be paid as reward to the MS
+    /// @param monitor_address Address of the MS
+    function claimReward(
+        address token_network_address,
+        uint192 reward_amount,
+        address monitor_address,
+        address closing_participant,
+        address non_closing_participant
+    )
+        public
+        returns (bool)
+    {
+        TokenNetwork token_network = TokenNetwork(token_network_address);
+        bytes32 channel_identifier = token_network.getChannelIdentifier(closing_participant, non_closing_participant);
+
+        // Only allowed to claim, if channel is settled
+        // Channel is settled if it's data has been deleted
+        uint256 channel_settle_block;
+        (, channel_settle_block, ) = token_network.getChannelInfo(closing_participant, non_closing_participant);
+        require(channel_settle_block == 0);
+
+        Reward storage reward = rewards[channel_identifier];
+
+        // Make sure that the Reward exists
+        require(reward.reward_sender_address != 0x0);
+
+        address raiden_node_address = recoverAddressFromRewardProof(
+            channel_identifier,
+            reward_amount,
+            token_network_address,
+            token_network.chain_id(),
+            reward.nonce,
+            monitor_address,
+            reward.reward_proof_signature
+        );
+
+        // Deduct reward from raiden_node deposit
+        balances[raiden_node_address] -= reward_amount;
+        // Add reward to the monitoring services' balance.
+        // This minimizes the amount of gas cost
+        // Only use token.transfer in the withdraw function
+        balances[monitor_address] += reward_amount;
+
+        emit RewardClaimed(monitor_address, reward_amount, channel_identifier);
+
+        // delete storage
+        delete rewards[channel_identifier];
+    }
+
+    /// @notice Withdraw depositted tokens. If msg.sender is MS don't allow to
+    /// withdraw below minimum_deposit.
+    /// Can be called by addresses with a deposit as long as they have a positive balance
+    /// @param amount Amount of tokens to be withdrawn
+    function withdraw(uint amount) public {
```

Comment:
For clarity, I would separate this into 2 functions for MS and for user (Raiden node).